transport/nio/src/test/java/org/kaazing/gateway/transport/tcp/TcpAcceptorIT.java

```diff
@@ -0,0 +1,271 @@
+        ConnectFuture x = connector.connect(bindAddress, handler, null);
+        x.addListener(new IoFutureListener(){
+
+            @Override
+            public void operationComplete(IoFuture arg0) {
+                latch.countDown();
+            }
+            
+        });
+        boolean didConnect = latch.await(1, SECONDS);
+        Assert.assertTrue("Fail to connect", didConnect);
+    }
+
+    @After
+    public void after() throws Exception {
+        // Make sure we always stop all I/O worker threads
+        if (connector != null) {
+            //schedulerProvider.shutdownNow();
+            connector.dispose();
+        }
+    }
+    
+    @Test
+    @Specification({
+        "establish.connection/tcp.server"
+        })
+    public void establishConnection() throws Exception {
+        connectTo8080(new IoHandlerAdapter());
+        k3po.finish();
+    }
+
+    @Test
+    @Specification({
+        "server.sent.data/tcp.server"
+        })
+    public void serverSentData() throws Exception {
+        connectTo8080(new IoHandlerAdapter());
+        
+        k3po.finish();
+    }
+
+    @Test
+    @Specification({
+        "client.sent.data/tcp.server"
+        })
+    public void clientSentData() throws Exception {
+        connectTo8080(new IoHandlerAdapter(){
+            @Override
+            protected void doSessionOpened(IoSession session) throws Exception {
+                ByteBuffer data = ByteBuffer.allocate(20);
+                String str = "client data";
+                data.put(str.getBytes());
+                
+                data.flip();
+
+                IoBufferAllocatorEx<?> allocator = ((IoSessionEx) session).getBufferAllocator();
+                
+                session.write(allocator.wrap(data.duplicate(), IoBufferEx.FLAG_SHARED));
+            }
+        });
+        
+        k3po.finish();
+    }
+
+    @Test
+    @Specification({
+        "bidirectional.data/tcp.server"
+        })
+    public void bidirectionalData() throws Exception {
+        connectTo8080(new IoHandlerAdapter(){
+            private int counter = 1;
+            
+            @Override
+            protected void doSessionOpened(IoSession session) throws Exception {
+                ByteBuffer data = ByteBuffer.allocate(20);
+                String str = "client data " + counter;
+                data.put(str.getBytes());
+                
+                data.flip();
+
+                IoBufferAllocatorEx<?> allocator = ((IoSessionEx) session).getBufferAllocator();
+                
+                session.write(allocator.wrap(data.duplicate(), IoBufferEx.FLAG_SHARED));
+            }
+            
+            @Override
+            protected void doMessageReceived(IoSession session, Object message) throws Exception {
+                String decoded = new String(((IoBuffer) message).array());
+
+                if (decoded.equals("server data " + counter) && counter < 2) {
```

Comment:
The data may be fragmented, for example, it may come as "server d"  followed by "ata 1". This is perfectly valid, so the test code needs to accommodate this.